### PR TITLE
fix: don't infer GQA present length from a symbolic past (#32716)

### PR DESCRIPTION
**Pull request by ykhrustalev**
**Problem**
`GroupQueryAttention` shape inference copies a symbolic past sequence length onto present, although present is `max(past, total_sequence_length)` and grows every step. The allocation planner compares `dim_param`s by name, so it gives present a past-sized buffer and the run fails with `Shape mismatch attempting to re-use buffer`. This happens through the WebGPU/CUDA `MayInplace` alias, and through the planner's free list on any EP, CPU included. Separately, a fused Conv `activation` the WebGPU EP cannot parse failed session creation with a bare failed-expression message.

**Solution**
- Leaves the present sequence dimension dynamic when the past one is a `dim_param`; a fixed-length past still propagates, so static caches are planned as before
- Keeps the WebGPU `MayInplace` declarations, which are only honoured when valid once the shape is right
- Leaves `SparseAttention` (`use_max_past_present_buffer == 1`) and `PagedAttention` (own inference) untouched
- Reports an unparseable fused Conv activation by name and points at core/EP version skew; drops the redundant parse in `FusedConv`

**Testing**
- New `GroupQueryAttentionGrowingCacheTest` asserts the inferred shape, runs the free-list route on CPU against `enable_mem_reuse = false`, and runs both routes on WebGPU while asserting GQA is assigned to the WebGPU EP; with only `bert_defs.cc` reverted, 4 of its 5 cases fail with the reported error
- Full `onnxruntime_provider_test` and `onnxruntime_test_all` pass on Metal; on Vulkan/lavapipe the only failures (`MathOpTest.Asin`, `CumSumTest._1DTestLong`, `ReductionOpTest.ArgMax_float_wide_last_axis_keepdims`, two `BatchedRightPaddedRotaryPrefillFlashAttention*_WebGPU`, `MatMulNBitsMlpFusion...WebGpuResults`) fail identically without this change
- Eight `onnx-community` GQA models (SmolLM2, Qwen2.5, gpt2, gemma-3, functiongemma, turn-detector, LFM2), CPU and WebGPU, growing and static caches, on both machines: logits bit-identical to the baseline wheel
- SmolLM2 and Qwen2.5 rewritten to a stacked, `Gather`-sliced KV cache: `Gather -> Where -> GQA` fails on WebGPU before and passes after, matching the unmodified model

Refs #32716

### CODE REVIEW

File: onnxruntime/test/contrib_ops/conv_unknown_activation_test.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+namespace test {
+
+// ConvActivationFusion writes the attribute from the core library while the activation set lives
+// in the EP, so a core library newer than the EP binary can fuse an activation the kernel has
+// never heard of. That has to report the activation by name, not a bare failed assertion.
+TEST(Conv_WebGPU, UnknownFusedActivationIsReportedByName) {
+  auto webgpu_ep = DefaultWebGpuExecutionProvider();
+  if (!webgpu_ep) {
+    GTEST_SKIP() << "WebGPU execution provider is not available.";
+  }
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply onnxruntime/test/contrib_ops/conv_unknown_activation_test.cc:25]: Fixed in e9e2274: this file moved to `test/contrib_ops/conv_unknown_activation_test.cc`. `filter_test_srcs` in `cmake/onnxruntime_unittests.cmake` drops the whole `test/contrib_ops` prefix when `onnxruntime_DISABLE_CONTRIB_OPS=ON`, so it is no longer compiled in that configuration and `FusedConv` is always available where it is.

File: onnxruntime/test/providers/webgpu/group_query_attention_growing_cache_test.cc [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+}  // namespace
+
+// The kernel def used to declare MayInplace(past_key -> present_key) unconditionally. Once the
+// planner honoured the alias, a step needing a longer present aborted in OpKernelContext::Output
+// with "Shape mismatch attempting to re-use buffer" -- every step of a grow-by-concat KV cache.
+TEST(GroupQueryAttention_WebGPU, PresentKvMayGrowBeyondPlannerOwnedPast) {
+  if (!DefaultWebGpuExecutionProvider()) {
+    GTEST_SKIP() << "WebGPU execution provider is not available.";
+  }
```
[the remark this answers is not in the document]
Comment (ykhrustalev) [reply onnxruntime/test/providers/webgpu/group_query_attention_growing_cache_test.cc]: Fixed in e9e2274: this file moved to `test/contrib_ops/group_query_attention_growing_cache_test.cc`, which `filter_test_srcs` excludes when `onnxruntime_DISABLE_CONTRIB_OPS=ON`. The test also grew to cover the shape inference directly and the free-list reuse route on the CPU EP, so it no longer depends on WebGPU being present.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 3]
```diff
@@ -357,6 +357,11 @@ void BaseGroupQueryAttentionTypeAndShapeInference(ONNX_NAMESPACE::InferenceConte
             // If total_sequence_length is provided and past_key has 0 length, present_key will grow.
             // Leave the dimension as dynamic to avoid "Error merging shape info" warning.
             present_shape.add_dim();
+          } else if (past_dims[2].has_dim_param()) {
```
Comment (qjia7) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc]: Please cover fixed positive past dimensions here as well. `present_length = max(past_length, total_sequence_length)`, so a numeric past dimension does not prove that present has the same length. For example, past `[1, 2, 16, 32]` with runtime `seqlens_k=[16]` and `total_sequence_length=17` requires present `[1, 2, 17, 32]`. This condition is false and the final `else` still infers length 16, allowing the same undersized-buffer reuse.

I reproduced this unchanged numeric-dimension path on `onnxruntime==1.31.0.dev20260920001` using `Gather -> Mul -> GroupQueryAttention -> Unsqueeze` on CPU: fixed past lengths 1, 2 and 16 with totals 2, 3 and 17 all succeed with memory reuse disabled and fail with it enabled. The 16-to-17 error is `Shape mismatch attempting to re-use buffer. {1,2,16,32} != {1,2,17,32}`. This is a remaining pre-existing case, rather than a new regression from this PR.

In the `use_max_past_present_buffer == -1` path, please replace the entire fallback for an unresolved maximum with `present_shape.add_dim()`, regardless of whether the past dimension is symbolic or numeric. Keep the exact maximum calculation when both values are known, the separate explicit fixed/windowed-cache branch, and the existing `MayInplace` declarations.

File: onnxruntime/test/contrib_ops/group_query_attention_growing_cache_test.cc [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+// A fixed-length past is a static cache shared with present, so its length still carries over.
+TEST(GroupQueryAttentionGrowingCacheTest, StaticPastLengthIsInferredForPresent) {
+  constexpr int64_t kMaxLen = 16;
+  std::shared_ptr<Model> model;
+  ASSERT_NO_FATAL_FAILURE(LoadModel(BuildKvCacheModel(PastSizedBuffer::kPastItself, Dim{kMaxLen}), model));
+
+  const auto* present_shape = InferredPresentKeyShape(model->MainGraph());
+  ASSERT_NE(present_shape, nullptr);
+  ASSERT_EQ(present_shape->dim_size(), 4);
+  EXPECT_EQ(present_shape->dim(2).dim_value(), kMaxLen);
```
Comment (qjia7) [inline onnxruntime/test/contrib_ops/group_query_attention_growing_cache_test.cc]: A fixed input dimension alone does not establish a shared static-cache contract. This model still has runtime `total_seq_len`, so feeding 17 with `kMaxLen=16` is valid and requires a present length of 17. The assertion currently locks in the incorrect inference described in the other comment.

Please change this test to expect an unknown present sequence dimension when total length is runtime-dependent, and add execution cases with fixed declared past lengths growing 1 -> 2, 2 -> 3 and 16 -> 17. The CPU free-list route should compare reuse enabled against reuse disabled; the WebGPU route should continue checking GQA assignment. Preserve precise-shape coverage separately for known constant total lengths and explicit fixed/windowed-cache semantics.

**Review by qjia7 — CHANGES_REQUESTED** [review]
Please split this into two PRs. Keep #32717 focused on the GQA shape-inference fix and its regression tests, and move the Conv/FusedConv activation-diagnostic changes and `conv_unknown_activation_test.cc` to a separate PR. These address independent problems and should be reviewed and merged separately.

The GQA change addresses the symbolic-past case, but the unresolved present length must also stay dynamic for a fixed positive past dimension when `total_sequence_length` is a runtime input. Please address the inline comments on the inference branch and the static-length test.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 3 above)
Comment (ykhrustalev) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: Done in dda193b: the `use_max_past_present_buffer == -1` fallback is now a single `present_shape.add_dim()`, so a fixed past length is left dynamic too. The exact `max(past, total)` for a constant total and the `sliding_window_cache` branch are unchanged, and the `MayInplace` declarations stay.

Your `Gather -> Mul -> GQA -> Unsqueeze` case with fixed past 1, 2 and 16 (totals 2, 3, 17) fails on the previous head with memory reuse enabled and passes with it, outputs identical to reuse disabled, on both Metal and Vulkan/lavapipe builds.

File: onnxruntime/test/contrib_ops/group_query_attention_growing_cache_test.cc (hunk 4 above)
Comment (ykhrustalev) [reply onnxruntime/test/contrib_ops/group_query_attention_growing_cache_test.cc]: Changed in dda193b. `FixedPastLengthLeavesPresentDynamic` now expects an unknown present length for a fixed past with a runtime total, and the CPU free-list test (reuse on vs off) and both WebGPU tests (with the GQA assignment check) run fixed declared pasts growing 1 -> 2, 2 -> 3 and 16 -> 17 on top of the symbolic steps. Exact-shape coverage is kept separately: `ConstantTotalLengthGivesExactPresentLength` (total as an initializer, 17 with past 16 -> 17, 8 with past 16 -> 16) and `SlidingWindowCacheKeepsPastLength` (`sliding_window_cache=1` keeps the capacity).

With only `bert_defs.cc` reverted, the two dynamic-shape tests and the three execution tests fail with `Shape mismatch attempting to re-use buffer`; the two exact-shape tests still pass.

**Comment by ykhrustalev** [comment]
Split as requested: the Conv/FusedConv activation diagnostics and `conv_unknown_activation_test.cc` moved to #32775 (71762a8 removes them here), so this PR is now `bert_defs.cc` plus `group_query_attention_growing_cache_test.cc`. The fixed-past case is addressed in dda193b; details in the inline replies.
